SimplenoteTests/NoticeContorllerTests.swift

```diff
@@ -1,33 +1,171 @@
+        XCTAssertEqual(controller.pendingNotices, 1)
     }
 
-    override func tearDownWithError() throws {
-        // Put teardown code here. This method is called after the invocation of each test method in the class.
+    func testMakeNoticeView() throws {
+        let notice = Notice(message: "Message", action: nil)
+
+        controller.present(notice)
+
+        let view = try XCTUnwrap(presenter.noticeView as? NoticeView)
+        XCTAssertEqual(notice.message, view.message)
+        XCTAssertNil(view.handler)
     }
 
-    func testExample() throws {
-        // This is an example of a functional test case.
-        // Use XCTAssert and related functions to verify your tests produce the correct results.
+    func testMakeNoticeViewWithAction() throws {
+        let action = NoticeAction(title: "Title") {
+            print("Action")
+        }
+        let notice = Notice(message: "Message ", action: action)
+
+        controller.present(notice)
+
+        let view = try XCTUnwrap(presenter.noticeView as? NoticeView)
+        XCTAssertEqual(notice.message, view.message)
+        XCTAssertEqual(notice.action?.title, view.actionTitle)
+        XCTAssertNotNil(view.handler)
     }
 
-    func testPerformanceExample() throws {
-        // This is an example of a performance test case.
-        self.measure {
-            // Put the code you want to measure the time of here.
+    func testMakeNoticeViewWithEmptyNotice() throws {
+        let action = NoticeAction(title: "") {
+            //No action
         }
+        let notice = Notice(message: "", action: action)
+
+        controller.present(notice)
+
+        let view = try XCTUnwrap(presenter.noticeView as? NoticeView)
+        XCTAssertEqual(notice.message, view.message)
+        XCTAssertNotEqual(notice.action?.title, view.actionTitle)
+        XCTAssertNotNil(view.handler)
     }
 
+    func testDismiss() {
+        let noticeA = Notice(message: "Message A", action: nil)
+
+        controller.present(noticeA)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+    }
+
+    func testDismissContinuesToNextNotice() {
+        let noticeA = Notice(message: "Message A", action: nil)
+        let noticeB = Notice(message: "Message B", action: nil)
+        controller.present(noticeA)
+        controller.present(noticeB)
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 1)
+
+        controller.dismiss()
+
+        XCTAssertTrue(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+
+        controller.dismiss()
+
+        XCTAssertTrue(presenter.dismissed)
+        XCTAssertFalse(controller.isPresenting)
+        XCTAssertEqual(controller.pendingNotices, 0)
+        XCTAssertEqual(presenter.actionLog, "presentNoticeView(Message A), dismissNotification(), presentNoticeView(Message B), dismissNotification(), ")
+    }
```

Comment:
An example how actionLog and mock timer can be used (I didn't run the actual test :-) )
```
func testDismissContinuesToNextNotice() {
    let noticeA = Notice(message: "Message A", action: nil)
    let noticeB = Notice(message: "Message B", action: nil)
    controller.present(noticeA)
    controller.present(noticeB)

    var expectedActions: [MockNoticePresenter.Action] = [
        .present(noticeA.message)
    ]
    XCTAssertEqual(presenter.actionLog, expectedActions)

    timerFactory.timer?.fire()

    expectedActions.append(.dismiss(noticeA.message))
    expectedActions.append(.present(noticeB.message))
    XCTAssertEqual(presenter.actionLog, expectedActions)

    timerFactory.timer?.fire()

    expectedActions.append(.dismiss(noticeB.message))
    XCTAssertEqual(presenter.actionLog, expectedActions)
}
```

This way we mimic how the actual flow would work as the client code wouldn't call `dismiss`.

We also don't need `controller.isPresenting` and `controller.pendingNotices`. And `controller.dismiss` can remain private.